### PR TITLE
[Hotfix] MAL imports

### PR DESCRIPTION
**Pull request by RaitoBezarius**
- Fix doctor's fields and use MALEntry.
- Refactor the whole MAL module using a `requests.Session` and instantiating an MALClient which handles the API operations.
- Rewrite the DB interaction to be more efficient and fix it in some edge cases.
- MALClient respects MAL by enforcing the `MAL_USER_AGENT` and does not uses `random_ip` (which is unnecessary according to my tests).
- Added basic tests.

The problem was: many errors due to inexistent animes, forbidden requests, and so on.

Solves a recent issue encountered by a user.

### CODE REVIEW

**Comment by jilljenn** [comment]
Sorry, cannot check now (requires a lot of attention :)). Maybe tomorrow.

**Comment by RaitoBezarius** [comment]
Don't worry!

**Comment by jilljenn** [comment]
`NameError: name 'hastattr' is not defined` :'( You could have tried before pushing ;P

**Comment by jilljenn** [comment]
(How could test possibly pass?)

File: mangaki/mangaki/templates/index.html [hunk 1]
```diff
@@ -20,9 +20,11 @@ <h3 style="margin-bottom: 1em">Notez vos séries favorites, on vous en fera déc
         <iframe src="https://ghbtns.com/github-btn.html?user=mangaki&repo=mangaki&type=fork&count=true&size=large" frameborder="0" scrolling="0" width="158px" height="30px"></iframe>
```
Comment (jilljenn) [inline mangaki/mangaki/templates/index.html]: Nit: &amp;

File: mangaki/mangaki/utils/mal.py [hunk 2]
```diff
@@ -31,149 +40,256 @@ def random_ip():
...
+        xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
+
+        return MALEntry(ET.fromstring(xml).find('entry'), work_type)
+
+    def search_works(self, work_type: MALWorks, query: str) -> List[MALEntry]:
+        resp = self._search_api(work_type, query)
+        self._translate_http_exceptions(resp)
+
+        html_code = html.unescape(re.sub(r'&amp;([A-Za-z]+);', r'a\1;', resp.text))
+        xml = re.sub(r'&([^alg])', r'&amp;\1', _encoding_translation(html_code))
```
Comment (jilljenn) [inline mangaki/mangaki/utils/mal.py]: Please don't duplicate code.

File: mangaki/mangaki/utils/mal.py [hunk 3]
```diff
@@ -31,149 +40,256 @@ def random_ip():
...
+    # All entries are supposed of the same type.
+    work_cat = Category.objects.get(slug=mal_entries[0].work_type.value)
+    for entry in mal_entries:
+        is_present = Work.objects.filter(
+            category=work_cat,
+            ext_poster=entry.poster
+        ).exists()
+
+        if not is_present:
+            Work.objects.create(
```
Comment (jilljenn) [inline mangaki/mangaki/utils/mal.py]: Discussion: I'm pretty sure this will add even more duplicates :'( Forever. Doushio. Use Reference ID?

File: mangaki/mangaki/utils/mal.py [hunk 4]
```diff
@@ -31,149 +40,256 @@ def random_ip():
...
 
 
-def import_mal(mal_username, mangaki_username):
+def get_or_create_from_mal(work_list: QuerySet,
+                           work_type: MALWorks,
+                           title: str,
+                           poster_link: str) -> Optional[Work]:
+    """
+    Get a work from the current `work_list` (a queryset, filtered by category)
+    or create from scratch using MAL as reference.
```
Comment (jilljenn) [inline mangaki/mangaki/utils/mal.py]: Please create an issue calling for a `Reference` check. (Add reference if does not exist.)

**Comment by RaitoBezarius** [comment]
Thanks for hotfixing my typo... We need (better) tests.

Regarding your comments, I am going to address them. I was not exactly sure
how I I was supposed to handle dupes. Thanks for the hint.

File: mangaki/mangaki/templates/index.html (hunk 1 above)
Comment (RaitoBezarius) [inline mangaki/mangaki/templates/index.html]: I didn't add it :p.

File: mangaki/mangaki/utils/mal.py (hunk 2 above)
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/mal.py]: Right, I put the work to deduplicate the code, but… forgot the last pass :D. Thanks!

File: mangaki/mangaki/utils/mal.py (hunk 3 above)
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/mal.py]: I'll add `Reference` objects.

File: mangaki/mangaki/utils/mal.py [hunk 5]
```diff
@@ -1,17 +1,16 @@
 """
-This module contains multiple functions to scrap MyAnimeList,
+This module contains multiple functions to interact with MyAnimeList API,
```
Comment (jilljenn) [inline mangaki/mangaki/utils/mal.py]: J'aime beaucoup :D
